[rendering] use default luminance for tone mapping of streams with bad metadata

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

Modification of the GetLuminanceValue() function:
- 600 nits for missing metadata (consistency with default for Reinhard that yields the hardcoded 0.7 value - instead of 400)
- 600 nits if the mastering display metadata is bad (.5 nits in sample linked in context, rounded to 0 because of integer calculations of the function)

I can only build/test Windows but noticed that GL/GLES have identical functions and made the change there as well.

Open questions, since I'm only aware of the one sample file:
* what's the max value that can be considered bad metadata? 0 is obviously bad metadata, but we could use something like 80 nits (SDR level) as the valid minimum, or some other value.
* what should the default value be? Reinhard defaults to 600, which worked better for the sample than the 400 already present in GetLuminanceValue(). For that sample, much higher values work even better to avoid blowing out brighter areas, but that is a sample and they may  have cranked everything to 11. 1000 nits is common in valid metadata.
* test for a max valid value of 10000?

Since there is code duplication between the platforms (identical GetLuminanceValue() functions, identical default for Reinhard), I don't mind creating a common function / const default but don't know where to locate it. Please suggest.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Black screen with tone mappers ACES Filmic and Hable for the sample [here ](https://4kmedia.org/sony-bravia-uhd-hdr-4k-demo/).
It has bad mastering display metadata (min 0.1 max 0.5) and no light metadata.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 10, now plays the sample ok with tone mapping ACES Filmic or Hable on SDR screen or HDR screen with option "Use diaplay HDR capabilities" set to off.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
All tone mapping algorithms provide an output for files with bad metadata.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai commented:
> Black screen with tone mappers ACES Filmic and Hable for the sample [here ](https://4kmedia.org/sony-bravia-uhd-hdr-4k-demo/).

This video is broken for sure:  I calculate that it should have a max. lum of 5000 nits (to look good) and it has meaningless values even for the x and y color coordinates (primaries):

![bad_video](https://user-images.githubusercontent.com/58434170/229366928-a2255fa1-093a-4e55-a301-bf62e7a01002.png)

I wouldn't worry too much about this video.

Default 400 nits value was chose careful for videos of real life (Disney+, Netflix, etc.) that NOT have HDR10 metadata (is different not have metadata at have bad/broken metadata).

Why 400?   because these videos usually are mastered with low max nits value to better compatibility with "low cost" devices and to watch these videos in tablets and smartphones and bad TV's. If you chose high value by default, these videos they look very dark.

In the other hand a video mastered with high luminance value (3000 - 5000 nits) is expected has correct metadata too. Is mandatory in these cases because much (or all) devices needs metadata to adapt source luminance to max device capable luminance. It is actually very difficult for a device to be capable of reproducing more than 1000 nits, then the higher the luminance of the source, the more important the metadata.

So is assumed that:

- Videos without metadata has low max. luminance values (300 - 500 nits)
- Videos with high max. luminance has correct metadata.

CrystalP commented:
No problem. Shouldn't the default for Reinhard be adjusted to 400 as well then?

I'll look for a place to dump the metadata in debug log, that could come in handy for troubleshooting.

thexai commented:
> No problem. Shouldn't the default for Reinhard be adjusted to 400 as well then?

Why? is a different algo and at the time it was not touched much to maintain backward compatibility with previous Kodi versions (do not change the video output for people who have saved individual tone map param values in the DB/VideoLibrary).

> I'll look for a place to dump the metadata in debug log, that could come in handy for troubleshooting.

fine, also can be seen in "Debug Info Video" (Alt + O) 

![debug](https://user-images.githubusercontent.com/58434170/229378433-5ca2221b-1b86-4500-858d-5a0d532f6297.png)

CrystalP commented:
In order to move forward, modified to 400 nits and refactored the function a bit.
> no behavior change compared to master version except for the case of display metadata present but invalid.
> moved a few variables around / eliminated variables
> shared and protected the calculation of mastering display max luminance (must have display metadata, luminance metadata, denominator must not be zero) because having light metadata is not a guarantee that display metadata is also there and valid.

Only applied to the Windows code for now, if you're OK I'll copy to GL and GLES.

thexai commented:
Changes in WinVideoFilter looks fine to me. 
Seems you applied different changes to GL and GLES... Is for some reason or is not finished yet?

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/YUV2RGBShaderGL.cpp:
```diff
@@ -258,13 +258,19 @@ void BaseYUV2RGBGLSLShader::SetToneMapParam(ETONEMAPMETHOD method, float param)
-  float lum2 = lum1;
-  float lum3 = lum1;
+  const float defaultLuminance =
+      400.0f; // default for bad quality HDR-PQ sources (with no metadata)
```
Please move comment on top to avoid ugly formatting

CrystalP committed:
[rendering] Create CToneMappers to share luminance calculation across platforms

CrystalP committed:
[rendering] use default luminance for tone mapping of streams with bad metadata

CrystalP commented:
Yes GL and GLES were not finished, but they are now. I copy/pasted from Windows.

Sharing the code would be good, I didn't see an obvious existing file for it though, see second commit for quick & dirty approach.
To be refactored in the future with calculations for Reinhard and proper class members / setters to avoid the ugly static method with many arguments.

thexai commented:
Runtime tested with some representative videos and confirmed calculated values still the same that in master:

```
The Arrival  Max ML 1000 / Max CLL 9977  -->  4067.40
Interstellar Max ML 4000 / Max CLL 1242  -->  1551.80
Free Guy   Max ML 1000 / Max CLL 0       -->  1000
The Mandalorian  (no metadata)           -->  400
```

thexai reviewed (approved):
Nice improvement have this in common code

enen92 commented:
this was merged while having the milestone set to beta 2

CrystalP commented:
oops thanks for fixing.
